### PR TITLE
Add logging util lib for Proto Message Logging (PML) http filter

### PR DESCRIPTION
**Pull request by dchakarwarti**
Commit Message: Add logging util lib for Proto Message Logging (PML) http filter
Additional Description: This functionality will be used in the PML filter
Risk Level: Low
Testing: Added unit tests
Docs Changes: The proto for PML filter is present at api/envoy/extensions/filters/http/proto_message_logging/v3/config.proto
Release Notes: Nil
Platform Specific Features: Nil

### CODE REVIEW

**Comment by dchakarwarti** [comment]
@moderation

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:1019]:
```
scorecard --repo github.com/grpc-ecosystem/proto-field-extraction
RESULTS
-------
Aggregate score: 3.1 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Branch-Protection      | branch protection not enabled  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#branch-protection      |
|         |                        | on development/release         |                                                                                                                       |
|         |                        | branches                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CI-Tests               | 0 out of 6 merged PRs          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                                                       |
|         |                        | normalized to 0                |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no effort to earn an OpenSSF   | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#cii-best-practices     |
|         |                        | best practices badge detected  |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 6 / 10  | Code-Review            | found 3 unreviewed changesets  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#code-review            |
|         |                        | out of 9 -- score normalized   |                                                                                                                       |
|         |                        | to 6                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Contributors           | 0 different organizations      | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#contributors           |
|         |                        | found -- score normalized to 0 |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Dangerous-Workflow     | no workflows found             | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dangerous-workflow     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Maintained             | 0 commit(s) out of 11 and 0    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#maintained             |
|         |                        | issue activity out of 0 found  |                                                                                                                       |
|         |                        | in the last 90 days -- score   |                                                                                                                       |
|         |                        | normalized to 0                |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Pinned-Dependencies    | no dependencies found          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#pinned-dependencies    |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                                                       |
|         |                        | 0                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Token-Permissions      | no github tokens found         | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#token-permissions      |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:1034]:
```
scorecard --repo github.com/grpc-ecosystem/proto_processing_lib
RESULTS
-------
Aggregate score: 3.4 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Branch-Protection      | branch protection not enabled  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#branch-protection      |
|         |                        | on development/release         |                                                                                                                       |
|         |                        | branches                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | CI-Tests               | no pull request found          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#ci-tests               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no effort to earn an OpenSSF   | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#cii-best-practices     |
|         |                        | best practices badge detected  |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 1 / 10  | Code-Review            | found 9 unreviewed changesets  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#code-review            |
|         |                        | out of 11 -- score normalized  |                                                                                                                       |
|         |                        | to 1                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 3 / 10  | Contributors           | 1 different organizations      | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#contributors           |
|         |                        | found -- score normalized to 3 |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Dangerous-Workflow     | no workflows found             | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dangerous-workflow     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 5 / 10  | Maintained             | 7 commit(s) out of 11 and 0    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#maintained             |
|         |                        | issue activity out of 0 found  |                                                                                                                       |
|         |                        | in the last 90 days -- score   |                                                                                                                       |
|         |                        | normalized to 5                |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Pinned-Dependencies    | no dependencies found          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#pinned-dependencies    |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | no SAST tool detected          | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#sast                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Token-Permissions      | no github tokens found         | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#token-permissions      |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/49c0eed3a423f00c872b5c3c9f1bbca9e8aae799/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:1004]: Scorecard score - [4.8](https://scorecard.dev/viewer/?uri=github.com/google/or-tools)

**Comment by nezdolik** [comment]
@dchakarwarti have you followed the [extension policy](https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#adding-new-extensions)? Does this extension have a maintainer sponsor?

File: bazel/repository_locations.bzl
Comment (kyessenov) [reply bazel/repository_locations.bzl:1004]: What's the purpose of combinatorial optimization for the extension? I'd be suspicious of using any algorithms here since they are typically super-linear.

**Comment by dchakarwarti** [comment]
> @dchakarwarti have you followed the [extension policy](https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#adding-new-extensions)? Does this extension have a maintainer sponsor?

Yes, we have followed the extension policy. The config proto for the PML filter has already been submitted previously: https://github.com/envoyproxy/envoy/pull/31735.


@yanavlasov is our sponsor.

File: bazel/repository_locations.bzl
Comment (kyessenov) [reply bazel/repository_locations.bzl:1004]: /wait-any

or-tools is research software, can you explain what it's used for?

Comment (dchakarwarti) [reply bazel/repository_locations.bzl:1004]: Hi Kuat!

`or-tools` isn't directly used in the logging libraries added for the PML filter. However, it is required for the Proto Processing Libraries. [Reference](https://github.com/grpc-ecosystem/proto_processing_lib/blob/09e6cf5da15c8b9d32c6706b2517890200110bc8/proto_processing_lib/proto_scrubber/BUILD.bazel#L94C11-L94C29).

The library is using `ortools/base/map_util.h`.

In envoy, the build fails if I do not explicitly import the `or-tools` library. 
```
$ bazel build //source/extensions/filters/http/proto_message_logging/logging_util:logging_util
ERROR: /usr/local/google/home/dchakarwarti/.cache/bazel/_bazel_dchakarwarti/0a1ddbe361fc4585fd11ce5863817141/external/com_google_protoprocessinglib/proto_processing_lib/proto_scrubber/BUILD.bazel:73:11: no such package '@com_google_ortools//ortools/base': The repository '@com_google_ortools' could not be resolved: Repository '@com_google_ortools' is not defined and referenced by '@com_google_protoprocessinglib//proto_processing_lib/proto_scrubber:field_mask_node'
ERROR: Analysis of target '//source/extensions/filters/http/proto_message_logging/logging_util:logging_util' failed; build aborted: 
INFO: Elapsed time: 0.865s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (25 packages loaded, 840 targets configured)
```

**Comment by soulxu** [comment]
@dchakarwarti it seems the code still have format issue.

**Comment by dchakarwarti** [comment]
> Details

Thanks Alex! Handled the formatting related issues.

File: bazel/repository_locations.bzl
Comment (kyessenov) [reply bazel/repository_locations.bzl:999]: I think it was added because it's coming from Google internal gtl, and or-tools exported some fork of it at some point?  Can you re-work your dependency not to depend on a repo to use gtl helper - you can simply copy paste those two helpers directly into your code.

Comment (kyessenov) [reply bazel/repository_locations.bzl:999]: A side question for these three proto libraries - are you somehow hooked into the internals of protobuf? Is there some hidden friend access? We can't update protobufs nearly as fast in OSS, so it'd be challenging if you use private proto interfaces.

Comment (dchakarwarti) [reply bazel/repository_locations.bzl]: >  you can simply copy paste those two helpers directly into your code.

Sure, I removed the `ortools` dependency from here.

Additionally, I also removed `@com_google_proto_field_extraction` external repo dependency as it was already being used as `@com_google_protofieldextraction`.

Now, the only newly introduced dependency is of `@com_google_protoprocessinglib`.

>  are you somehow hooked into the internals of protobuf?

Actually, no. There is no hidden friend access. 

To add more context, this library is part of [grpc-ecosystem](https://github.com/grpc-ecosystem/proto_processing_lib) repo from which we have been using the field extraction and proto converter libraries as well.

**Comment by dchakarwarti** [comment]
cc: @yanavlasov

**Review by yanavlasov — DISMISSED** [review]

**Comment by yanavlasov** [comment]
LGTM from me. Please merge main to fix the coverage build.

I will wait for @kyessenov approval and then merge.

**Comment by dchakarwarti** [comment]
> LGTM from me. Please merge main to fix the coverage build.
> 
> I will wait for @kyessenov approval and then merge.

Sure, addressed the coverage build.

Added some more unit tests. Requesting a review please.

Thank you!

**Comment by dchakarwarti** [comment]
@kyessenov can you please take a look at the dependency changes?

**Review by kyessenov — APPROVED** [review]
Approving deps change - don't have the full context for this filter. Thanks for addressing the extraneous import.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
